### PR TITLE
BACK-1617: Integrate Concentrated arUSD/fxProtocol rUSD

### PR DESCRIPTION
**Pull request by JackieJoo**

### CODE REVIEW

**Comment by Wertusser** [comment]
https://linear.app/paraswap/issue/BACK-1626

**Review by Wertusser — APPROVED** [review]

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    );
+
+    return [
+      {
+        unit: unitOut,
+        prices: amountsOut,
+        data: {},
+        poolAddresses: [this.config.rUSDAddress],
+        exchange: this.dexKey,
+        gasCost: 1000000,
```
Comment (0xNazarii) [inline src/dex/fx-protocol-rusd/fx-protocol-rusd.ts:197]: 1 mil seems too high, lets look at examples and use the actual cost

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+      this.config.rUSDWeETHMarketAddress,
+      this.rUSDMarketIface,
+      this.config.weETHOracleAddress,
+      this.weETHOracleIface,
+      this.logger,
+    );
+  }
+
+  async initializePricing(blockNumber: number) {
+    const data: { returnData: any[] } =
```
[the remark this answers is not in the document]
Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts]: fixed

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<FxProtocolData>> {
+    if (side == SwapSide.BUY) {
```
[the remark this answers is not in the document]
Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts]: fixed

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    ).toString();
+
+    const weETHPrice = BigInt(
+      this.weETHOracleIface.decodeFunctionResult(
+        'latestAnswer',
+        data.returnData[1],
+      )[0],
+    ).toString();
+
+    await Promise.all([
```
[the remark this answers is not in the document]
Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts:89]: fixed

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts (hunk 1 above)
Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts:167]: yes, its a actual gas cost, on average one swap take 960000-1115220 gas

Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts:167]: same for weETH<>rUSD and rUSD<>arUSD swaps

File: src/dex/fx-protocol-rusd/fx-protocol-rusd.ts [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    const is_rUSD_src = this.is_rUSD(srcToken);
+    const is_weETH_src = this.is_weETH(srcToken);
+    const is_rUSD_dest = this.is_rUSD(destToken);
+    const is_weETH_dest = this.is_weETH(destToken);
+
+    if (is_weETH_src && is_rUSD_dest) {
+      const exchangeData = this.fxUSDIface.encodeFunctionData('mint', [
+        this.config.weETHAddress,
+        srcAmount,
+        this.dexHelper.config.data.executorsAddresses?.Executor01,
```
Comment (0xNazarii) [inline src/dex/fx-protocol-rusd/fx-protocol-rusd.ts:216]: doesnt this needs to be `recipient`?

Comment (Wertusser) [reply src/dex/fx-protocol-rusd/fx-protocol-rusd.ts]: fixed for `mint` and `redeem`

**Review by 0xNazarii — APPROVED** [review]

**Pull request merged**